engine/flink/components/base-tests/src/test/scala/pl/touk/nussknacker/engine/flink/ResultCollectingListenerSpec.scala

```diff
@@ -0,0 +1,337 @@
+          Map("input" -> 10),
+          Map("input" -> 20),
+          Map("input" -> 30),
+          Map("input" -> 40),
+        )
+        transitionVariables(testResults, "sub", Some("sub-filter")) shouldBe Set(
+          Map("fragment1_input" -> 10),
+          Map("fragment1_input" -> 20),
+          Map("fragment1_input" -> 30),
+          Map("fragment1_input" -> 40),
+        )
+        transitionVariables(testResults, "sub", Some("sub-filter")) shouldBe Set(
+          Map("fragment1_input" -> 10),
+          Map("fragment1_input" -> 20),
+          Map("fragment1_input" -> 30),
+          Map("fragment1_input" -> 40),
+        )
+        // This sample is filtered out and does not proceed further
+        transitionVariables(testResults, "sub-filter", None) shouldBe Set(
+          Map("fragment1_input" -> 10),
+        )
+        transitionVariables(testResults, "sub-filter", Some("sub-fragmentEnd")) shouldBe Set(
+          Map("fragment1_input" -> 20),
+          Map("fragment1_input" -> 30),
+          Map("fragment1_input" -> 40),
+        )
+        transitionVariables(testResults, "sub-fragmentEnd", Some("end")) shouldBe Set(
+          Map("input" -> 20, "fragmentResult" -> Map("output" -> 20)),
+          Map("input" -> 30, "fragmentResult" -> Map("output" -> 30)),
+          Map("input" -> 40, "fragmentResult" -> Map("output" -> 40)),
+        )
+        transitionVariables(testResults, "end", None) shouldBe Set(
+          Map("input" -> 20, "fragmentResult" -> Map("output" -> 20)),
+          Map("input" -> 30, "fragmentResult" -> Map("output" -> 30)),
+          Map("input" -> 40, "fragmentResult" -> Map("output" -> 40)),
+        )
+      }
+    )
+  }
+
+  private def transitionVariables(testResults: TestResults[Any], fromNodeId: String, toNodeId: Option[String]) =
+    testResults
+      .nodeTransitionResults(NodeTransition(fromNodeId, toNodeId))
+      .map(_.variables)
+      .toSet[Map[String, Any]]
+      .map(_.map { case (key, value) => (key, scalaMap(value)) })
+
+  private def scalaMap(value: Any): Any = {
+    value match {
+      case hashMap: java.util.HashMap[_, _] => hashMap.asScala.toMap
+      case other                            => other
+    }
+  }
+
+  private def assertNumberOfSamplesThatFinishedInNode(testResults: TestResults[Any], sinkId: String, expected: Int) =
+    testResults.nodeTransitionResults.get(NodeTransition(sinkId, None)).map(_.length) shouldBe Some(expected)
+
+  private def withCollectingTestResults(
+      canonicalProcess: CanonicalProcess,
+      assertions: TestResults[Any] => Unit
+  ): Unit = {
+    ResultsCollectingListenerHolder.withListener { collectingListener =>
+      val model = modelData(collectingListener)
+      flinkMiniCluster.withDetachedStreamExecutionEnvironment { env =>
+        val executionResult = new FlinkScenarioUnitTestJob(model).run(canonicalProcess, env)
+        flinkMiniCluster.withRunningJob(executionResult.getJobID) {
+          eventually {
```

Comment:
Let's avoid using eventually. It can introduce some strange situations like flaky tests, false positives etc. Our test cases are (correct me if I'm wrong) deterministic. We can follow the approach:
```
flinkMiniCluster.waitForJobIsFinished(executionResult.getJobID)
assertions(collectingListener.results)
```